lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -4899,6 +4894,65 @@ void ConformanceChecker::ensureRequirementsAreSatisfied() {
 
 #pragma mark Protocol conformance checking
 
+/// Determine whether mapping the interface type of the given protocol non-type
+/// requirement into the context of the given conformance produces a well-formed
+/// type.
+static bool
+hasInvalidTypeInConformance(const ValueDecl *requirement,
+                            NormalProtocolConformance *conformance) {
+  assert(isa<TypeDecl>(requirement) == false);
+  assert(requirement->getDeclContext()->getSelfProtocolDecl() ==
+         conformance->getProtocol());
+
+  // FIXME: getInterfaceType() on properties returns pseudocontextual types.
+  auto interfaceTy = requirement->getInterfaceType();
```

Comment:
What's a pseudocontextual type?

Reply:
I provided an example in the tests. It is a type that was first mapped into context, resulting in some type parameters being replaced with concrete types vs. archetypes, and then mapped out of context with these concrete types still in place. This is the reason `getInterfaceType()` on a `var` returns `Never` instead of `Self.A` when the latter is constrained to `Never`.

Reply:
Ah, ok. "pseudocontextual" is not a word we use elsewhere so it might be best to provide an explanation here.

If you want to fix that inconsistency once and for all, take a look at the code in TypeCheckPattern. The interface type of a VarDecl comes from its Pattern. We should refactor Patterns to use interface types, at least if they appear in non-local context. Since you eliminated TypeResolution::forContextual(), I think what's happening is there are matching mapTypeIntoContext()/mapTypeOutOfContext() calls somewhere that just needs to be removed.

Reply:
> Ah, ok. "pseudocontextual" is not a word we use elsewhere so it might be best to provide an explanation here.

Agreed, will add a comment here too.